### PR TITLE
Extract OPML-handling code from controller

### PR DESCRIPTION
**Pull request by Minoru**
(This PR is part of a decomposition effort; see #213)

This moves the meat of OPML handling out of `controller` and into newly-created `OPML` class. `controller` now handles the boring I/O details, which means `OPML` is more testeable. The new class comes with a full test coverage, of course.

We deal with OPML in two places: `controller`'s import-export, and `opml_urlreader`. I meant to combine them before submitting a PR, but got sidetracked and lost first the context and then the interest. So let's merge this first, and I'll file a separate issue to continue the work later.

@kmws, wanna review this?

### CODE REVIEW

File: test/opml_urlreader.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+		{"https://example.com/mirrors/distrowatch.rss", {}},
+		{"https://example.com/feed.atom", {}},
+		{"https://example.com/feed.rss09", {}},
+		{"https://blogs.example.com/~john/posts.rss", {"Blogs"}},
+		{"https://fred.example.com/writing/index.php?type=rss", {"eloquent"}},
+		{"https://blogs.example.com/~mike/.rss", {"Blogs/friends"}},
+		{"https://one.example.com/file.xml", {}},
+		{"https://to.example.com/elves/tidings.rss", {}},
+		{"https://third.example.com/~of/internet.rss", {"Rant boards/humans"}},
+        {"https://four.example.com/~john/posts.rss", {"Rant boards"}},
```
Comment (der-lyse) [inline test/opml_urlreader.cpp]: Looks like this indentation is broken.

File: src/opml.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+						"OPML import: url = %s is "
+						"already "
+						"in list",
+						url);
+				}
+				xmlFree(url);
+			} else {
+				char* text = (char*)xmlGetProp(
+					node, (const xmlChar*)"text");
+				if (!text)
```
Comment (der-lyse) [inline src/opml.cpp]: I miss the curly braces. :-)

File: test/opml_urlreader.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
```
Comment (der-lyse) [inline test/opml_urlreader.cpp]: Same indentation issue here.

File: include/file_urlreader.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#ifndef NEWSBOAT_FILE_URLREADER_H_
+#define NEWSBOAT_FILE_URLREADER_H_
+
+#include <string>
+
+#include "urlreader.h"
+
+namespace newsboat {
+
+class file_urlreader : public urlreader {
```
Comment (der-lyse) [inline include/file_urlreader.h]: Didn't you want to use UpperCamelCase for classes in new code? Technically speaking a refactoring produces new code. :-P But this class existed before, so the class itself is not new, yeah.

File: src/opml.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
+#include "opml.h"
+
+#include <cassert>
+#include <cstring>
+
+namespace newsboat {
+
+xmlDocPtr OPML::prepare_opml(const FeedContainer& feedcontainer)
```
Comment (der-lyse) [inline src/opml.cpp]: The term "prepare" sounds a tiny bit odd to me, does "generate" fit better?

File: src/opml.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	if (doc == nullptr) {
+		return false;
+	}
+
+	xmlNode* root = xmlDocGetRootElement(doc);
+
+	for (xmlNode* node = root->children; node != nullptr;
+		node = node->next) {
+		if (strcmp((const char*)node->name, "body") == 0) {
+			LOG(level::DEBUG, "import_opml: found body");
```
Comment (der-lyse) [inline src/opml.cpp]: Should be just `import` instead of `import_opml`.

File: src/opml_urlreader.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			"opml_urlreader::reload: downloading `%s'",
+			url);
+		std::string urlcontent = utils::retrieve_url(url, cfg);
+
+		xmlDoc* doc =
+			xmlParseMemory(urlcontent.c_str(), urlcontent.length());
+
+		if (doc == nullptr) {
+			LOG(level::ERROR,
+				"opml_urlreader::reload: parsing XML file "
```
Comment (der-lyse) [inline src/opml_urlreader.cpp]: Does it make sense to include the URL a well? The log above containing the URL is just of debug level.

File: src/opml_urlreader.cpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+			std::string theurl(rssurl);
+			urls.push_back(theurl);
+			if (tag.length() > 0) {
+				std::vector<std::string> tmptags;
+				tmptags.push_back(tag);
+				tags[theurl] = tmptags;
+				alltags.insert(tag);
+			}
+		}
+		if (rssurl)
```
Comment (der-lyse) [inline src/opml_urlreader.cpp]: Curly braces are missing.

File: test/opml.cpp [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+		REQUIRE(check(feeds) == expectedOpmlText);
+	}
+
+	SECTION("A few feeds") {
+		configcontainer cfg;
+		cache rsscache(":memory:", &cfg);
+		FeedContainer feeds;
+
+		std::shared_ptr<rss_feed> feed
+			= std::make_shared<rss_feed>(&rsscache);
```
Comment (der-lyse) [inline test/opml.cpp]: Other code segments have the `=` in the line above. Not sure, whether we are free to choose or our style guide wants us to do it a specific way.

**Comment by der-lyse** [comment]
Apart from the minor things I discovered it's looking good to me.

Off-topic, but I have to say when reading the heavily indented code in the OMPL class, that the line length limit is way too low in my opinion. Having so many statements split over multiple lines does not make the code more readable to me, the opposite is true. I don't mind deeply nested blocks, but they  don't make a nice combo with a hard limit of at most 80 characters. E.g. a couple of times I just thought the code was broken and then after a few seconds noticed that the `for` loop or `if` continued in the next line (the indentation of the continued line just looked like regular block code). Well, it's probably just a matter of getting used to.

File: include/file_urlreader.h (hunk 4 above)
Comment (kmws) [reply include/file_urlreader.h]: I guess the rationale is that in this step we only extract this class from the file. During this step the name should not be changed. The same way you don't change the code during transition -- that would tell why the curly braces are missing for one-liners.

File: test/opml.cpp [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+using namespace newsboat;
+
+TEST_CASE("prepare_opml creates an XML document with feed URLs in OPML format",
+		"[opml]")
+{
+	const auto check =
+		[]
+		(const FeedContainer& feedcontainer)
+		-> std::string
```
Comment (kmws) [inline test/opml.cpp]: Does clang-format format it like that? feels like a waste. 3 lines that could be one and wouldn't lose readability.

**Comment by kmws** [comment]
Looks good to me, besides little formatting that @der-lyse mentioned. I think adding the test suite for this is huge improvement -- Kudos!

File: src/opml.cpp (hunk 5 above)
Comment (Minoru) [reply src/opml.cpp]: I think it does; I'll update the code. Thanks!

File: test/opml.cpp (hunk 9 above)
Comment (Minoru) [reply test/opml.cpp]: The style guide doesn't specify. I'll move `=` to the first line for consistency.

File: include/file_urlreader.h (hunk 4 above)
Comment (Minoru) [reply include/file_urlreader.h]: Yeah, this was just me moving things around, not even refactoring.

CameCase issue was never resolved, actually. That's a mistake, I'll got fix it; thanks for the nudge! :)

**Comment by Minoru** [comment]
@der-lyse, I agree with you. The proper way to fix this, in my opinion, is to indent less and refactor into smaller functions. In this PR, I didn't do that because I was simply moving some existing code to a new place. I hope whoever takes up #292 will do refactor it into a better shape.

@kmws, I sent you an invite to become a collaborator on this repository. You *can*, but don't *have to*, tag issues, close them (I guess), and push small commits with typo fixes of whatnot directly to mater. Please submit a PR for anything more substantial than that.

Thanks for the reviews, both of you! I fixed everything you spotted, and intend to merge this when Travis greenlights the changes.

**Pull request merged**